### ISSUE TITLE
Tiny hole in the mesh of VoxelLodTerrain

### ISSUE THREAD

**Issue by Piratux** [Zylann/godot_voxel#641]
**Describe the bug**
As title says, I found a case where the meshed SDF of VoxelLodTerrain contains tiny hole.

**To Reproduce**
Download project and run it. Upon launch you should see this which contains triangle shaped hole:

![image](https://github.com/Zylann/godot_voxel/assets/58703216/b2055751-f232-4c99-b162-a23aa4183d6b)

**Expected behavior**
There should be no holes in the mesh.

**Environment**
- OS: Windows 10
- Graphics card: GTX 650 ti
- Versions: Godot 4.2.2 stable + voxel tools 1.2.0
- Renderer: Forward+

[Test Mesh issue.zip](https://github.com/user-attachments/files/15523024/Test.Mesh.issue.zip)

**Comment by Zylann** [comment]
It appears to be caused by this: https://github.com/Zylann/godot_voxel/blob/4187c63df62496e7e1a1d08fe05c516b36a49050/meshers/transvoxel/transvoxel.cpp#L712-L749

The mesher eliminates triangles that are too small or degenerate, because that makes Jolt Physics spam errors.
Maybe there can be a way to turn this off, but then if you use Jolt you'll end up with the problem again.

**Comment by Piratux** [comment]
I see.
This is not very noticeable so it's not a big problem in my case.
Just wanted to let you know.

Feel free to close the issue.

**Comment by Zylann** [comment]
Another option would be to snap/clamp edge interpolation factors so that vertices cannot be generated too close to corners, which then tends to produce less thin/small triangles.
This post mentions a similar idea: https://stackoverflow.com/questions/35112181/marching-cubes-very-small-triangles (unfortunately without any code example so it's unclear what they mean by "snapping iso values")

**Comment by Piratux** [comment]
From the description and provided images in the post of your link, it sounds like if mesh vertices are very close to integer coordinates (which are corners of the cubes), then vertices get set to those integer coordinate.

Something like this I'd assume:

```c
float epsilon = 0.05;
if (pos.x * pos.x + pos.y * pos.y + pos.z * pos.z < epsilon * epsilon){
    pos.x = std::round(pos.x);
    pos.y = std::round(pos.y);
    pos.z = std::round(pos.z);
}
```

**Comment by Zylann** [comment]
No, you're only describing the result of it. Vertices are generated on edges of marching cube cells, and that depends on an interpolation factor between pairs of corners. That factor can be clamped/snapped, but the SDF values can also be clamped/snapped. I'm not sure which one they are referring to. One downside of doing it on the SDF is that it requires to know how gradients progress, and depends on them being good gradients, while the interpolation factor has no unit and so is easier to clamp (however it is known after case selection so triangle count won't change)

**Comment by Piratux** [comment]
Considering the author of post there mentioned percentages, I can only imagine this working with the interpolation factor since it can be a fixed range if [0, 1], where 0 is starting point of the edge, and 1 is ending point.

If SDF can be "infinitely large", then I don't see how percentages can be used to "snap" the SDF.

**Comment by Zylann** [comment]
I removed the code that was removing small triangles, replaced it with a margin property that does edge interpolation clamping.
